### PR TITLE
feat: App Proxy provisioning and workflows

### PR DESCRIPTION
**Pull request by jachym-tousek-keboola**
Jira: https://keboola.atlassian.net/browse/PSGO-368

**Changes:**
- Added provisioning and workflows for app proxy service.

-----------

You can skip the first commit during review I think. It's a straight copy of templates-api provisioning. All adjustments specific to app proxy are in later commits.

### CODE REVIEW

File: provisioning/app-proxy/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+Directories:
+- `dev` - various auxiliary files for development.
+- `docker` - docker image of the service.
+- `kubernetes` - templates of the Kubernetes configurations.
+
+Included files (they are not called directly):
+- `common.sh` - main file for all deployments.
+- `aws.sh` - deploy script for AWS stacks.
+- `azure.sh` - deploy script for Azure stacks.
+- `local.sh` - deploy script for a local MiniKube deployment.
```
Comment (michaljurecko) [inline provisioning/app-proxy/README.md:19]: :heavy_plus_sign: `gcp.sh`

File: provisioning/app-proxy/azure.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  --query "properties.outputs.clusterName.value" \
+  --output tsv)
+
+# Get credentials to the Azure Managed Kubernetes Service
+az aks get-credentials --name "$CLUSTER_NAME" --resource-group "$RESOURCE_GROUP" --overwrite-existing
+
+# Common part of the deploy
+export ETCD_STORAGE_CLASS_NAME=
+. ./common.sh
+
```
Comment (michaljurecko) [inline provisioning/app-proxy/azure.sh:27]: There is no etcd.

File: provisioning/app-proxy/kubernetes/templates/cloud/aws/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  name: app-proxy
+  namespace: $NAMESPACE
+  labels:
+    app: app-proxy
+  annotations:
+    kubernetes.io/ingress.class: "nginx"
+spec:
+  rules:
+    - host: "hub.${HOSTNAME_SUFFIX}"
+      http:
```
Comment (michaljurecko) [inline provisioning/app-proxy/kubernetes/templates/cloud/aws/ingress.yaml:14]: There should be two records, `hub....`, `*.hub.`, see: https://kubernetes.io/docs/concepts/services-networking/ingress/#hostname-wildcards

File: provisioning/app-proxy/kubernetes/templates/cloud/gcp/ingress.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  name: app-proxy
+  namespace: $NAMESPACE
+  labels:
+    app: app-proxy
+  annotations:
+    kubernetes.io/ingress.class: "nginx"
+spec:
+  rules:
+    - host: "hub.${HOSTNAME_SUFFIX}"
+      http:
```
Comment (michaljurecko) [inline provisioning/app-proxy/kubernetes/templates/cloud/gcp/ingress.yaml:14]: There should be two records, `hub....`, `*.hub.`, see: https://kubernetes.io/docs/concepts/services-networking/ingress/#hostname-wildcards

File: provisioning/app-proxy/kubernetes/templates/proxy/config-map.yaml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: app-proxy
+  namespace: $NAMESPACE
+  labels:
+    app: app-proxy
+data:
+  sandboxesApiHost: "http://sandboxes-service-api.default.svc.cluster.local)"
```
Comment (michaljurecko) [inline provisioning/app-proxy/kubernetes/templates/proxy/config-map.yaml:10]: `.local)"` -> `.local"`

**Comment by michaljurecko** [comment]
FYI: OpenApi validator fix https://github.com/keboola/keboola-as-code/pull/1414/commits/159a9c4bc7a69f478bc249ecf31cb478c5445153

File: provisioning/app-proxy/common.sh [hunk 6]
```diff
@@ -24,6 +24,13 @@ export NAMESPACE="app-proxy"
 # Namespace
 kubectl apply -f ./kubernetes/deploy/namespace.yaml
 
+if ! kubectl get secret app-proxy --namespace app-proxy > /dev/null 2>&1; then
+  COOKIE_SECRET_SALT=$(head /dev/urandom | tr -dc A-Za-z0-9 | head -c 128)
+  export COOKIE_SECRET_SALT
+  envsubst < kubernetes/templates/proxy/secret.yaml > kubernetes/deploy/proxy/secret.yaml
+  kubectl apply -f ./kubernetes/deploy/proxy/secret.yaml
+fi
+
```
Comment (michaljurecko) [inline provisioning/app-proxy/common.sh:33]: :+1:

File: provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                  name: app-proxy
+                  key: cookieSecretSalt
+            - name: DD_AGENT_HOST
+              valueFrom:
+                fieldRef:
+                  fieldPath: status.hostIP
+            - name: DD_TRACE_ANALYTICS_ENABLED
+              value: "true"
+            - name: DD_PROFILING_ENABLED
+              value: "true"
```
Comment (kacurez) [inline provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:79]: tato env je tu imo zbytocna, podla docs to ani nezapina ten profiling. Potrebujete vlastne ten DD profiling (plati sa zato zvlast)

File: provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+              port: 8000
+              httpHeaders:
+                - name: Host
+                  value: KubernetesLivenessProbe
+            periodSeconds: 10
+            failureThreshold: 3
+      tolerations:
+        - key: app
+          operator: Exists
+          effect: NoSchedule
```
Comment (kacurez) [inline provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:123]: tie `tolerations` tu tiez uz nemusia byt

File: provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml (hunk 7 above)
Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:79]: ok :) to je asi 5 generacia, kde sa to iba kopiruje, neviem, ktora bola povodna service

Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:79]: Jachym, ked to budes vyhadzovat, tak to prosim vyhod z celeho repa.

File: provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml (hunk 8 above)
Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:123]: Tie tolerations teda mozeme tiez vyhodit z celeho repa.

Comment (kacurez) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml]: ano, ale radsej by som tu upravu/vyhodenie nespajal s tymto PR a poriesil to bokom v novom PR.

Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml]: nie je to ok :) je to genericka sablona, odstranit sa moze toto:
https://github.com/keboola/keboola-as-code/blob/eebcdc5872b0d54515e2ef815832c8f6b23d45fb/provisioning/common/etcd/values.yaml#L12-L15

File: provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            }
+          }
+    spec:
+      containers:
+        - name: app-proxy
+          image: $APP_PROXY_REPOSITORY:$APP_PROXY_IMAGE_TAG
+          resources:
+            requests:
+              cpu: "200m"
+              memory: "768Mi"
```
Comment (kacurez) [inline provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: memory request (tj pamat s ktorou pocita k8s scheduler pri alokacii podu na node) by som znizil, [vidim](https://app.datadoghq.eu/dashboard/gny-t8t-d99/kubernetes?fromUser=true&fullscreen_end_ts=1709031416093&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1708945016093&fullscreen_widget=6879185011415216&refresh_mode=sliding&tpl_var_kube_namespace%5B0%5D=app-proxy&view=spans&from_ts=1708945014095&to_ts=1709031414095&live=true) ze to realne zabera okolo 30MB, tak request na 50MB namiesto 768MB mi pride vhodnejsi, navyse limit je 768MB.

Comment (kacurez) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: ak sa naopak pocita s tym ze produkcny load vyzaduje prave tolko vela pamate, tj konstantne okolo 768MB, tak by som dvihol limit na 1GB.

Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: Este sme to neskusali pod zatazou, pod `500MB` (limit) by som nesiel. Staci nam to v 2 instanciach.
Vacsinou ste z SRE chceli, aby bol request a limit rovnaky. Ak sa to zmenilo, tak request samozrejme moze ist dolu.

Comment (kacurez) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: > Vacsinou ste z SRE chceli, aby bol request a limit rovnaky.

aha, to si nespominam, neviem si teraz predstavit kde by to mohlo davat zmysel okrem pripadu ked to je konstantne od zaciatku velmi vysoke. Request je to s cim pocita scheduler ked sklada pody na nody, limit je hranica pri ktorej terminuje pody. V [service requirements](https://keboola.atlassian.net/wiki/spaces/ENGG/pages/1453785161/Service+requirements#Resource-limits-%26-requests) je k tomu napisane ze request ma byt max 75% limit. Ak vam pride rozumny min limit 500MB tak request by som dal 150MB, alebo teda pri 768MB limite 200MB request. 
Kazdopadne mozeme to sledovat a postupne upravovat.

Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: :heavy_check_mark: 

ok, dava zmysel, je mi jasne ako to funguje, len mi nebolo jasne ake hodnoty tam chcete

> aha, to si nespominam, 

nejdem asi hladat PR starsie ako 1 rok, ale z toho dovodu sme davali vsade rovnake hodnoty

> V service requirements je k tomu napisane ze request ma byt max 75% limit. 

**[DOCS](https://keboola.atlassian.net/wiki/spaces/ENGG/pages/1453785161/Service+requirements#Resource-limits-%26-requests): Requesty by nemělý být míň než ~75% limitů, jinak se může stát, že instance bude busy a začne ukončovat kontejnery.** 

@kacurez toto nesedi ^^^ `nemělý být míň než ~75% limitů`, `500*0.75=375MB`, ty pises `150MB` - tj. to sedi s tou mojou predstavou, ze sme to davali rovnake (resp nie pod 75%).

Teda ako to ma byt? :slightly_smiling_face:

Comment (kacurez) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: j ja som to bral opacne a zle som si to vypocital, [tu je napr prehlad limitov a requestov api deploymentov](https://app.datadoghq.eu/dashboard/gny-t8t-d99/kubernetes?fromUser=false&fullscreen_end_ts=1709033792675&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1708947392675&fullscreen_widget=6678268604183702&refresh_mode=sliding&tpl_var_env%5B0%5D=kbc-testing-azure-east-us-2&tpl_var_kube_deployment%5B0%5D=%2Aapi&tpl_var_kube_namespace%5B0%5D=%2A&view=spans&from_ts=1708947302353&to_ts=1709033702353&live=true) - vacsina z nich ma ten pomer okolo 50% aj menej, tak teraz neviem odkial sa to cislo >75% berie :D 
image

Tiez napriklad [buffer-api](https://app.datadoghq.eu/dashboard/gny-t8t-d99/kubernetes?fromUser=false&fullscreen_end_ts=1709033866628&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1706441866628&fullscreen_widget=6879185011415216&refresh_mode=sliding&tpl_var_kube_deployment%5B0%5D=buffer-api&tpl_var_kube_namespace%5B0%5D=%2A&view=spans&from_ts=1708947302353&to_ts=1709033702353&live=true) (req = limit = 512MB) za posledny mesiac malo peak v US 470MB a EU 250MB ale v ostatnych stackoch je vacsinou pod 200MB. 
Ale ok, ak nevieme kolko to bude zrat pamate tak mozme to na zaciatok nechat vysoko.

Comment (michaljurecko) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: @kacurez a zhodneme sa na tom, ze si to vypocital inak ako je to v docs? :slightly_smiling_face: 

Mozeme tam dat kludne `150MB`, len to prosim fixnime/upravne aj v docs, ... nech nepride o mesiac niekto iny z SRE teamu a nepyta sa, "preco to nemate podla nasich standardov?".

Comment (kacurez) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: jj, tak dajte to prosim na 50% z limitu. To aky limit, necham na vas.
S tymi standardami som si nabehol no :) Poznacim si to bokom a preberiem s SRE.

Comment (Halama) [reply provisioning/app-proxy/kubernetes/templates/proxy/deployment.yaml:47]: prosel jsem si, doc je cajk. ne vsechno ji ale respektuje a ne vsude to dava smysl. nechte to takhle zatim.

**Review by michaljurecko — APPROVED** [review]

**Review by kacurez — APPROVED** [review]
na Azure testing to je nasadene a [potvrdene](https://keboolaglobal.slack.com/archives/C055WKPHYSD/p1709027888276049?thread_ts=1708958536.183119&cid=C055WKPHYSD) ze funguje.

**Pull request merged**